providers/azure.go

```diff
@@ -149,13 +145,76 @@ func (p *AzureProvider) Redeem(ctx context.Context, redirectURL, code string) (*
 	created := time.Now()
 	expires := time.Unix(jsonResponse.ExpiresOn, 0)
 
-	return &sessions.SessionState{
+	session := &sessions.SessionState{
 		AccessToken:  jsonResponse.AccessToken,
 		IDToken:      jsonResponse.IDToken,
 		CreatedAt:    &created,
 		ExpiresOn:    &expires,
 		RefreshToken: jsonResponse.RefreshToken,
-	}, nil
+	}
+
+	// the presence of id_token depends on the configuration and
+	// verification of id_token requires oidc configuration
+	if session.IDToken != "" && p.Verifier != nil {
+		email, err := p.verifyIDTokenAndExtractEmail(ctx, session.IDToken)
+		if err != nil {
+			return nil, err
+		}
+		session.Email = email
+	}
+
+	return session, nil
+}
+
+// EnrichSession finds the email to enrich the session state
+func (p *AzureProvider) EnrichSession(ctx context.Context, s *sessions.SessionState) error {
+	if s.Email != "" {
+		return nil
+	}
+
+	email, err := getEmailFromProfileAPI(ctx, s.AccessToken, p.ProfileURL.String())
+	if email == "" || err != nil {
+		err = fmt.Errorf("unable to get email address: %v", err)
```

Comment:
I think this will lead to misleading error message with an empty trailing `:` when `email == ""` and `err == nil`

Reply:
refactored a bit